src/missions/use-after-free-control-flow/temporal-mission.c

```diff
@@ -0,0 +1,295 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ * Copyright (c) 2020 Microsoft, Inc.
+ */
+#include <assert.h>
+#include <ctype.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <sys/queue.h>
+
+#if defined(__CHERI_PURE_CAPABILITY__)
+#include <sys/caprevoke.h>
+static void __attribute__((used)) *check_caprevoke = caprevoke;
+extern void malloc_revoke(void);
+static void __attribute__((used)) *check_malloc_revoke = malloc_revoke;
+#endif
+
+char *cstr[] = {
+	"Chelan", /* Pacific Northwest */
+	"Colt", /* United Kingdom */
+	"Morello" /* Poised to take over the world */
+};
+static const size_t ncstr = sizeof(cstr)/sizeof(cstr[0]);
+
+struct farm;
+struct crop;
+
+static void
+success(struct farm *f)
+{
+	fprintf(stderr, "Exploit successful: computer bought the farm!\n");
+	exit(42);
+}
+
+TAILQ_HEAD(cropq, crop);
+
+struct farm {
+	struct cropq	cropq;
+	struct crop	*cursor;
+	void		*pad;
+};
+
+struct crop {
+	void			(*describe)(struct crop *);
+	char			*type;
+	TAILQ_ENTRY(crop)	farm_cropq;
+};
+
+_Static_assert(sizeof(struct crop) == sizeof(struct farm),
+  "Structure size mismatch");
+
+union ufo {
+	void	*ptrA;
+	char	buf[sizeof(struct crop)];
+};
+
+_Static_assert(sizeof(struct crop) == sizeof(union ufo),
+  "Structure size mismatch");
+
+static void
+descr_1(struct crop *c)
+{
+	fprintf(stderr, "%s at %p\n", c->type, c);
+}
+
+static void
+descr_2(struct crop *c)
+{
+	fprintf(stderr, "%p holds a %s\n", c, c->type);
+}
+
+static unsigned int cid;
+static union ufo *ufo;
+
+static const size_t nfarm = 4;
+struct farm *farmp[nfarm];
+
+static void
+rm_farm(int fix)
+{
+	struct farm *f = farmp[fix];
+
+	farmp[fix] = NULL;
+	if (f != NULL) {
+		struct crop *c, *tc;
+
+		TAILQ_FOREACH_SAFE(c, &f->cropq, farm_cropq, tc) {
+			TAILQ_REMOVE(&f->cropq, c, farm_cropq);
+			free(c);
+		}
+
+		fprintf(stderr, "Tear down farm (index %d) at %p\n", fix, f);
+		free(f);
+	}
+}
+
+static struct farm *
+mk_farm(int fix)
+{
+	struct farm *f;
+
+	rm_farm(fix);
+
+	f = malloc(sizeof(struct farm));
+	assert(f != NULL); /* Surely infinite memory */
+
+	TAILQ_INIT(&f->cropq);
+	f->cursor = NULL;
+
+	farmp[fix] = f;
+
+	fprintf(stderr, "New farm (index %d) at %p\n", fix, f);
+	return f;
+}
+
+static void
+rm_crop(struct farm *f, struct crop *c)
+{
+	fprintf(stderr, "Del crop at %p\n", c);
+	TAILQ_REMOVE(&f->cropq, c, farm_cropq);
+	free(c);
+}
+
+static struct crop *
+mk_crop(struct farm *f)
+{
+	struct crop *c;
+
+	c = malloc(sizeof(struct crop));
+	assert(c != NULL);
+
+	if (f->cursor != NULL) {
+		/* Inherit description of current cursor */
+		c->describe = f->cursor->describe;
+		c->type = f->cursor->type;
+	} else {
+		c->describe = (cid & 1) ? descr_1 : descr_2 ;
```

Comment:
This is now entirely artificial. I preferred it when the function itself had the crop type in it, as that's more reminiscent of the classic OO interface + override examples you see (Animal interface has speak method, Dog implements that as printing bark, Cat implements that as printing meow).

Reply:
That's fair; I'll put it back.